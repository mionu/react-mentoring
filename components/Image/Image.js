import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core';

const noImagePath = '/no-image.png';

const useStyles = makeStyles({
    notVisible: {
        opacity: 0,
        position: 'absolute',
    },
    visible: {
        opacity: 1,
        position: 'relative',
    },
});

export default function Image(props) {
    const [loaded, setLoaded] = useState(false);
    const classes = useStyles();

    return <>
        <img
            className={loaded ? classes.visible : classes.notVisible}
            width={props.width}
            height={props.height}
            alt={props.alt}
            src={props.src}
            onLoad={() => setLoaded(true)}
            onError={() => setLoaded(false)}
        ></img>
        <img
            className={loaded ? classes.notVisible : classes.visible}
            width={props.width}
            height={props.height}
            alt={props.alt}
            src={noImagePath}
        ></img>
    </>;
}

Image.propTypes = {
    width: PropTypes.number,
    height: PropTypes.number,
    src: PropTypes.string,
    alt: PropTypes.string,
};

Image.defaultProps = {
    width: 300,
    height: 400,
};
