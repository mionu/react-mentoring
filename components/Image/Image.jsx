import React, { useState } from 'react';
import PropTypes from 'prop-types';
import useStyles from './styles';

export default function Image({
  alt, height, src, width,
}) {
  const [loaded, setLoaded] = useState(false);
  const classes = useStyles();

  return (
    <>
      <img
        className={loaded ? classes.visible : classes.notVisible}
        width={width}
        height={height}
        alt={alt}
        src={src}
        onLoad={() => setLoaded(true)}
        onError={() => setLoaded(false)}
      />
      <img
        className={`${classes.noImage} ${loaded ? classes.notVisible : classes.visible}`}
        width={width}
        height={height}
        alt={alt}
      />
    </>
  );
}

Image.propTypes = {
  alt: PropTypes.string,
  height: PropTypes.number,
  src: PropTypes.string.isRequired,
  width: PropTypes.number,
};

Image.defaultProps = {
  alt: '',
  height: 400,
  width: 300,
};
