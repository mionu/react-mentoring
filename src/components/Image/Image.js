import React from 'react';
import PropTypes from 'prop-types';
import NoImage from '../../../assets/no-image.png';

export default function Image(props) {
    const setFallbackImage = ({ target }) => {
        target.src = NoImage;
    }

    return <img
        width={props.width}
        height={props.height}
        alt={props.alt}
        src={props.src}
        onError={setFallbackImage}
    ></img>
}

Image.propTypes = {
    width: PropTypes.number,
    height: PropTypes.number,
    src: PropTypes.string,
    alt: PropTypes.string,
};

Image.defaultProps = {
    width: 300,
    height: 400,
};
