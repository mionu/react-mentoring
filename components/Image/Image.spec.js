import React from 'react';
import renderer, { act } from 'react-test-renderer';
import Image from './Image';

describe('Image', () => {
    it('should render correctly', () => {
        const tree = renderer
            .create(<Image src='http://asd.com' />)
            .toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('should accept width and height props', () => {
        const tree = renderer
            .create(<Image width={500} height={750} src='http://asd.com' />)
            .toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('should change visible image after loading', () => {
        const component = renderer.create(
            <Image src='http://asd.com' />
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        const img = component.root.findAllByType('img')[0];
        expect(img).toBeDefined();
        
        act(() => img.props.onLoad());

        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should keep fallback image if image not loaded', () => {
        const component = renderer.create(
            <Image src='http://asd.com' />
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        const img = component.root.findAllByType('img')[0];
        expect(img).toBeDefined();
        
        act(() => img.props.onError());

        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
