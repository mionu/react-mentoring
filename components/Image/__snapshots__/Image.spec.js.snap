// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should accept width and height props 1`] = `
Array [
  <img
    alt=""
    className="makeStyles-notVisible-2"
    height={750}
    onError={[Function]}
    onLoad={[Function]}
    src="http://asd.com"
    width={500}
  />,
  <img
    alt=""
    className="makeStyles-noImage-1 makeStyles-visible-3"
    height={750}
    width={500}
  />,
]
`;

exports[`Image should change visible image after loading 1`] = `
Array [
  <img
    alt=""
    className="makeStyles-notVisible-2"
    height={400}
    onError={[Function]}
    onLoad={[Function]}
    src="http://asd.com"
    width={300}
  />,
  <img
    alt=""
    className="makeStyles-noImage-1 makeStyles-visible-3"
    height={400}
    width={300}
  />,
]
`;

exports[`Image should change visible image after loading 2`] = `
Array [
  <img
    alt=""
    className="makeStyles-visible-3"
    height={400}
    onError={[Function]}
    onLoad={[Function]}
    src="http://asd.com"
    width={300}
  />,
  <img
    alt=""
    className="makeStyles-noImage-1 makeStyles-notVisible-2"
    height={400}
    width={300}
  />,
]
`;

exports[`Image should keep fallback image if image not loaded 1`] = `
Array [
  <img
    alt=""
    className="makeStyles-notVisible-2"
    height={400}
    onError={[Function]}
    onLoad={[Function]}
    src="http://asd.com"
    width={300}
  />,
  <img
    alt=""
    className="makeStyles-noImage-1 makeStyles-visible-3"
    height={400}
    width={300}
  />,
]
`;

exports[`Image should keep fallback image if image not loaded 2`] = `
Array [
  <img
    alt=""
    className="makeStyles-notVisible-2"
    height={400}
    onError={[Function]}
    onLoad={[Function]}
    src="http://asd.com"
    width={300}
  />,
  <img
    alt=""
    className="makeStyles-noImage-1 makeStyles-visible-3"
    height={400}
    width={300}
  />,
]
`;

exports[`Image should render correctly 1`] = `
Array [
  <img
    alt=""
    className="makeStyles-notVisible-2"
    height={400}
    onError={[Function]}
    onLoad={[Function]}
    src="http://asd.com"
    width={300}
  />,
  <img
    alt=""
    className="makeStyles-noImage-1 makeStyles-visible-3"
    height={400}
    width={300}
  />,
]
`;
